app/multitenant/consul_pipe_router.go

```diff
@@ -0,0 +1,605 @@
+			continue
+		}
+		if kvp != nil {
+			pipe = &consulPipe{}
+			if err := pipe.fromBytes(kvp.Value); err != nil {
+				log.Errorf("Error deserialising pipe %s: %v", key, err)
+				continue
+			}
+			index = kvp.ModifyIndex // if it doesn't exist, it will be 0
+		}
+
+		if pipe, retry, err = f(pipe); err != nil {
+			log.Errorf("Error CASing pipe %s: %v", key, err)
+			if !retry {
+				return nil, err
+			}
+			continue
+		}
+
+		if pipe == nil {
+			panic("Callback must instantiate pipe!")
+		}
+
+		value, err := pipe.toBytes()
+		if err != nil {
+			log.Errorf("Error serialising pipe %s: %v", key, err)
+			continue
+		}
+		ok, _, err := kv.CAS(&consul.KVPair{
+			Key:         key,
+			Value:       value,
+			ModifyIndex: index,
+		}, writeOptions)
+		if err != nil {
+			log.Errorf("Error CASing pipe %s: %v", key, err)
+			continue
+		}
+		if !ok {
+			log.Errorf("Error CASing pipe %s, trying again %d", key, index)
+			continue
+		}
+		return pipe, nil
+	}
+	return nil, fmt.Errorf("Failed to aquire pipe")
+}
+
+// Watch a given pipe and trigger a callback when it changes.
+// if callback returns false or error, exit (with the error).
+func (pr *consulPipeRouter) watch(key string, deadline time.Time, f func(*consulPipe) (bool, error)) (*consulPipe, error) {
+	var (
+		index = uint64(0)
+		kv    = pr.client.KV()
+	)
+	for deadline.After(mtime.Now()) {
+		// Poll waiting for the entry to get updated
+		kvp, meta, err := kv.Get(key, &consul.QueryOptions{
+			RequireConsistent: true,
+			WaitIndex:         index,
+			WaitTime:          longPollDuration,
+		})
+		if err != nil {
+			return nil, fmt.Errorf("Error getting %s: %v", key, err)
+		}
+		if kvp == nil {
+			return nil, fmt.Errorf("Pipe %s unexpectedly deleted!", key)
+		}
+
+		pipe := &consulPipe{}
+		pipe.fromBytes(kvp.Value)
+		if ok, err := f(pipe); !ok {
+			return pipe, nil
+		} else if err != nil {
+			return pipe, err
+		}
+
+		index = meta.LastIndex
+	}
+	return nil, fmt.Errorf("Timed out waiting on %s", key)
+}
+
+// negotiate tries to ensure the given end of the given pipe
+// is 'owned' by this pipe service replica in consul.
+func (pr *consulPipeRouter) negotiate(key string, e app.End) (*consulPipe, error) {
+	// First try and establish a record with us owning one end
+	_, err := pr.cas(key, func(p *consulPipe) (*consulPipe, bool, error) {
+		if p == nil {
+			p = &consulPipe{
+				CreatedAt: mtime.Now(),
+			}
+		}
+		if !p.DeletedAt.IsZero() {
+			return nil, false, fmt.Errorf("Pipe %s has been deleted", key)
+		}
+		end := p.end(e)
+		if end != "" && end != pr.advertise {
+			return nil, true, fmt.Errorf("Error: Pipe %s has existing connection to %s", key, end)
+		}
+		p.setEnd(e, pr.advertise)
+		p.incr(e)
+		return p, false, nil
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	// Next wait for other end to connect
+	// at this point we 'own' one end
+	pipe, err := pr.watch(key, mtime.Now().Add(10*longPollDuration), func(p *consulPipe) (bool, error) {
```

Comment:
Its consul's job to serialise all these.  Our app1.watch can't miss  the app2.cas - if we do the watch after the cas, we will see the new value (as we start with version 0), otherwise we will see the value when the version changes.

Does that make sense?